standard/types.md

```diff
@@ -705,3 +714,182 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+
+> *Note:* When the nullable annotation context is disabled, `T` and `T?` both refer to the same types. *end note*
+
+### §Nullable-Contexts Nullable context
+
+#### §Nullable-Contexts-General General
+
+Every line of source code has a ***nullable context***. The nullable context controls whether nullable annotations (§Nullable-Annotation-Context) have effect and whether nullable warnings (§Nullable-Warning-Context) are issued by the compiler. The nullable context can be one of *disabled*, *enabled*, *annotations*, or *warnings*. The *enabled* setting combines both *annotations* and *warnings*.
+
+The nullable context may be specified within source code via nullable directives (§Nullable-Directives) and/or via some implementation-specific mechanism external to the source code. If both approaches are used, nullable directives supersede the settings made via an external mechanism.
+
+The default state of the nullable context is implementation defined.
+
+Throughout this specification, all C# code that does not contain nullable directives, or about which no statement is made regarding the current nullable context state, shall be assumed to have been compiled with nullable context enabled.
+
+#### §Nullable-Disable-Context Nullable disable
+
+When the nullable context is ***disabled***:
+
+- A variable of any reference type is non-nullable.
+- No warning shall be generated when a variable of an unannotated reference type is initialized with, or assigned a value of, `null`.
+- No warning shall be generated when a variable of a reference type that possibly has the null value.
+- For any reference type `T`, the annotation `?` in `T?` is ignored. An informational message should be generated to that effect.
+  > *Note*: This message is characterized as “informational” rather than “warning,” so as not to confuse it with the state of the nullable warning setting, which is unrelated. *end note*
+- The null-forgiving operator `!` is ignored.
+
+> *Example*:
+>
+> <!-- Example: {template:"code-in-main-without-using", name:"NullableAnnotationContext1", ignoredWarnings:["CS0219","CS8632"], expectedException:"NullReferenceException"} -->
+> ```csharp
+> #nullable disable
+> string? s1 = null;    // Informational message; ? is ignored
+> string s2 = null;     // OK; null initialization of a reference
+> s2 = null;            // OK; null assignment to a reference
+> char c1 = s2[1];      // OK; no warning on dereference of a possible null; throws NullReferenceException
+> c1 = s2![1];          // OK; ! is ignored
+> ```
+>
+> *end example*
+
+#### §Nullable-Annotation-Context Nullable annotations
```

Comment:
Anchoring this here somewhat randomly... the annotation aspect affects "question marks in this bit of code" not "whether or not the compiler understands that a type may be nullable". Example:

```csharp
#nullable enable
string? M2() => null;

#nullable disable annotations
// Warning, because this code still "knows" that M2 returns string?, even if we couldn't declare a variable that way.
// (But we can declare it with var...)
Console.WriteLine(M2().ToString());
```